static/js/user_groups.ts

```diff
@@ -153,3 +153,26 @@ export function get_user_groups_of_user(user_id: number): UserGroup[] {
     );
     return groups_of_user;
 }
+
+export function get_recursive_subgroups(user_group_id: number): Set<number> | undefined {
+    const user_group = user_group_by_id_dict.get(user_group_id);
+    if (user_group === undefined) {
+        blueslip.error(`Could not find user group with ID ${user_group_id}`);
+        return undefined;
+    }
+
+    const subgroup_ids = new Set(user_group.subgroups);
+
+    for (const subgroup_id of subgroup_ids) {
+        const user_group = user_group_by_id_dict.get(subgroup_id);
+        if (user_group === undefined) {
+            blueslip.error(`Could not find user group with ID ${subgroup_id}`);
+            return undefined;
+        }
+
+        for (const group_id of user_group.subgroups) {
```

Comment:
I don't like the variable names of `user_group_id`, `subgroup_id`, `group_id` here.  I think best would be:
* `direct_subgroup_id` for this loop variable.
* `subgroup_id` for the outer loop variable is probably fine, but then the `const user_group` variable should be `const subgroup`, I think.
* `user_group_id` is OK at the top, but maybe something like `target_group_id` might feel better.